### ISSUE TITLE
Digest macro uses arbitrary prefixes

### ISSUE THREAD

**Issue by tcharding** [comit-network/comit-rs#2624]
## Problem 

Currently we use arbitrary prefixes for fields in the digestable struct
```
#[derive(Clone, Digest, Debug, PartialEq)]
#[digest(hash = "SwapDigest")]
pub struct HanEtherereumHalightBitcoinCreateSwapParams {
	...
    #[digest(prefix = "2001")]
    pub ethereum_absolute_expiry: Timestamp,
    #[digest(prefix = "2002")]
    pub ethereum_amount: asset::Ether,
	...
```

## Goal

Make sense of these prefixes, also design what they should be as we add more structs.  Also of note do we require a digestable struct for each swap or just a single one for a portocol pair (since the fields are the same)?

**Comment by bonomat** [comment]
@thomaseizinger / @tcharding : 
What do you think of taking a few bytes of a sha256 of the field's name as the digest? 
It is unlikely to get a clash here as each struct can have particular field name only once.

**Comment by thomaseizinger** [comment]
> @thomaseizinger / @tcharding :
> What do you think of taking a few bytes of a sha256 of the field's name as the digest?
> It is unlikely to get a clash here as each struct can have particular field name only once.

While that would work for making it unique, it goes against the original idea of the digest.
Eventually, I would want these prefixes to be documented in the RFCs/specs so anyone implementing a compatible protocol can just copy them from there.
That was also the reason we made those configurable in the first place :)

And easy way forward would be to come up with a list of fields that should go into the digest and then just generate random bytes for each of them and document them somewhere.

**Comment by bonomat** [comment]
I looked into the spike and couldn't find any details on this. Can you summarize what the purpose is please (if it is more than what's written in the [code](https://github.com/comit-network/comit-rs/blob/01e37bb24983c5944fea758943c5841908ae2706/digest/src/lib.rs#L3-L16))? 

From my current understanding a part of an hash would just work perfect and would reduce the documentation to a minimum :) 

> And easy way forward would be to come up with a list of fields that should go into the digest and then just generate random bytes for each of them and document them somewhere.

If we go for random bytes, why not leave it as it is?

**Comment by thomaseizinger** [comment]
> I looked into the spike and couldn't find any details on this. Can you summarize what the purpose is please (if it is more than what's written in the [code](https://github.com/comit-network/comit-rs/blob/01e37bb24983c5944fea758943c5841908ae2706/digest/src/lib.rs#L3-L16))?

Yeah I think that was not super well-documented back then :sweat_smile: 

> From my current understanding a part of an hash would just work perfect and would reduce the documentation to a minimum :)

It would work but it would be conceptually wrong. The name of the field (and hence the hash) should not be the source of truth. The source of truth should be the protocol specification (which doesn't exist at this stage).

Every implementation of the announce protocol that wants to be compatible has to be able to control _exactly_, which bytes go into the hash of each individual value that is part of the digest.
If we make this implicit, several things can go wrong:

1. A simple rename would be a breaking change on the public API.
2. I am not sure if the string encoding of a field name is stable across different Rust versions (I think it is UTF-8 but I am not sure). The same codebase compiled with different versions of Rust could be incompatible.

Even if (2) is guaranteed to not happen, I wouldn't want to depend on implementation details like this for a public API.
The string encoding is certainly not consistent across programming languages. C doesn't use UTF-8.

This means, regardless of the approach that we choose to come up with a prefix, be it hashing of the field name or specifying something statically, we have to document these bytes so other implementations can use the exact same and be compatible.

If we have to document them explicitly anyway, we should also specify them statically otherwise we have to somehow test that the internals we rely on don't get out of sync with the documentation. Seems like an unnecessary burden to do that.

Relying on internals like the string encoding of field names reminds me of technologies like Java's RMI. Yes you can write software in another programming language that is compatible with it but it is hard and brittle.
From my understanding, we are striving for protocols that are easy to implement in all kinds of contexts.

> If we go for random bytes, why not leave it as it is?

The current values suggest that there is a pattern / system on how these should be named, yet there is no documentation on that. If we decide to go random, I would go full random (similar to Bitcoin's magic bytes per network).

**Comment by bonomat** [comment]
Thanks for the nice summary, totally worth it to have this reasoning written down. Why the hash wouldn't work makes sense to me now.


My understanding is that a particular field (e.g. `pub bitcoin_expiry: Timestamp` should have the same prefix across different SwapParams). Is this correct?, e.g. like this: 

```
pub struct SwapWithFields {
    #[digest(prefix = "2001")]
	pub some_field_that_does_not_matter: foo,

	// bitcoin types
    #[digest(prefix = "3001")]
    pub bitcoin_expiry: Timestamp,
    #[digest(prefix = "3002")]
    pub bitcoin_amount: asset::Bitcoin,
}
```

```
pub struct DifferentSwapWithDifferentFields {
	// same bitcoin types
    #[digest(prefix = "3001")]
    pub bitcoin_expiry: Timestamp,
    #[digest(prefix = "3002")]
    pub bitcoin_amount: asset::Bitcoin,
   	
	// some other types
    #[digest(prefix = "4001")]
	pub some_field_that_does_not_matter_neither: bar,
}
```

**Comment by thomaseizinger** [comment]
> My understanding is that a particular field (e.g. `pub bitcoin_expiry: Timestamp` should have the same prefix across different SwapParams). Is this correct?, e.g. like this:

Yes, the value should be tied to the semantics of the field.
Note that, at the moment, we don't yet add all the fields that I originally envisioned.
To make the digest more powerful, I think we should also include:

- The PeerIds of each role (i.e. Alice = Qm123123...., Bob = Qm163463464...)
- The network of each ledger
- The protocol we are going to use on each ledger

If you choose to have a unique prefix for each value, i.e. you have a prefix for `bitcoin_expiry`, then we will also need to include an enum that says, which ledger is on which side, i.e. alpha-ledger = bitcoin (otherwise, I think,  swaps with completely flipped parameters will have the same digest). Alternatively, we make prefixes for the generic names like `alpha_expiry`. Not quite sure what the better approach is :grimacing: 

In the end, the goal is to have matching digests for the same swap and different ones for different swaps :)

**Comment by bonomat** [comment]
Oh, I did not see your response... when pushed out my PR.

>If you choose to have a unique prefix for each value, i.e. you have a prefix for bitcoin_expiry, then we will also need to include an enum that says, which ledger is on which side, i.e. alpha-ledger = bitcoin (otherwise, I think, swaps with completely flipped parameters will have the same digest).

I think this is unlikely to happen unless you use exactly the same values for the timestamps. i.e. two swaps with the same parameters just flipped directions at the same time sounds unlikely to me.


Nevertheless, it seems like there is a need for some more thinking: 

First I was in favor of the enum but I think this will limit us in the long run, e.g. same-chain swaps would not be possible. Also, right now I'm not happy with the naming:

```
pub struct HbitHerc20SwapParams {
    ...
    pub bitcoin_identity: identity::Bitcoin,
    pub bitcoin_expiry: Timestamp,
    pub bitcoin_amount: asset::Bitcoin,

    pub ethereum_identity: identity::Ethereum,
    pub ethereum_expiry: Timestamp,
    pub erc20_amount: asset::Erc20Quantity,
    pub token_contract: identity::Ethereum,
}
```

Without knowledge about what Hbit means, the field names won't mean anything to me. Given the struct is called `HbitHerc20...` I think the fields should be called something like `hbit_identity` to indicate that this is an identity relating to this protocol. 
However, if we would do this we would have: 

```
pub struct HbitHbitSwapParams {
    ...
    pub hbit_identity: identity::Bitcoin,
    pub hbit_expiry: Timestamp,
    pub hbit_amount: asset::Bitcoin,

    pub hbit_identity: identity::Bitcoin,
    pub hbit_expiry: Timestamp,
    pub hbit_amount: asset::Bitcoin,
}

```

... which would not work, so we would need to add some prefixes ala `alpha` and `beta`. I thought that this is out of scope for the swap digest but it might actually solve the problem: 


```
pub struct HbitHbitSwapParams {
    ...
    pub alpha_hbit_identity: identity::Bitcoin,
    pub alpha_hbit_expiry: Timestamp,
    pub alpha_hbit_amount: asset::Bitcoin,

    pub beta_hbit_identity: identity::Bitcoin,
    pub beta_hbit_expiry: Timestamp,
    pub beta_hbit_amount: asset::Bitcoin,
}

```

in this case `beta_hbit_identity` has the implicit meaning of being the `redeem_identity` from the perspective of Alice. 

With prefixes annotated we would have 
```
pub struct HbitHbitSwapParams {
    ...
    #[digest(prefix = "0001")]
    pub alpha_hbit_identity: identity::Bitcoin,

... [removed the rest] 

    #[digest(prefix = "0002")]
    pub beta_hbit_identity: identity::Bitcoin,
}

```

And when used in a different swap, e.g. 


```
pub struct HbitHerc20SwapParams {
    ...
    #[digest(prefix = "0001")]
    pub alpha_hbit_identity: identity::Bitcoin,

... [removed the rest] 

    #[digest(prefix = "0003")]
    pub beta_herc20_identity: identity::Ethereum,
}

```


--- 

So the question is, do we define the `prefix` by the semantical meaning of that field or the type of it?

**Comment by D4nte** [comment]
> Quick offline discussion led to this decision:
> 
> we are going for a mix between type specific and semantical prefixes. Additionally we will add another field specifying which protocol is used on each side and id's for bob and alice.
> Old

Hum, any chance you could describe what is the Problem/Goal? The Goal in the description is unclear to me.
